### ISSUE TITLE
BlockResults query can return incorrect info for older blocks

### ISSUE THREAD

**Issue by SpicyLemon** [cometbft/cometbft#3354]
## Bug Report

The `BlockResults` query can mistakenly think an entry is using the new `ResponseFinalizeBlock` when it was actually stored as a `LegacyABCIResponses` (had the name `ABCIResponses` when originally stored). When that happens, the content of the `ResponseFinalizeBlock` is weird and most of the data is just lost.

The effect is that we won't be able to reliably get historical block results from before our next upgrade.

### Setup

**CometBFT version** (use `cometbft version` or `git rev-parse --verify HEAD` if installed from source): `v0.38.7`

**Have you tried the latest version**: no (it just came out today but has the same `LoadFinalizeBlockResponse`).

**ABCI app** (name for built-in, URL for self-written if it's publicly available): provenanced

**Environment**:
- **OS** (e.g. from /etc/os-release): MacOS Sonoma 14.5
- **Install tools**:
- **Others**:

**node command runtime flags**:

### Config

(defaults)

### What happened?

We are testing a software upgrade that will bring us from `v0.46` of the SDK, and `v0.34.29` of CometBFT (via replace line), to `v0.50` of the SDK and `v0.38.7` of CometBFT. For these tests, we (manually) fire up a local one-node chain, submit a couple Txs, submit a software upgrade proposal, make the deposit and vote it through. Then, once the upgrade height is hit, stop that processes, switch to our new version, and start it up again. The upgrade runs as expected and the node starts cutting blocks again.

Then, as part of the checks, we ran a block-results query (on the updated node) for a height that was before the upgrade and had the gov-vote Tx in it (trying to see what the old begin and end block events look like now). The results came back very weird:

```json
{
  "jsonrpc": "2.0",
  "id": -1,
  "result": {
    "height": "8",
    "txs_results": [
      {
        "code": 0,
        "data": "CgoIgIDAChCAjs4cEg4IoI0GEgQIgMYKGICAQBoJCgdlZDI1NTE5",
        "log": "",
        "info": "",
        "gas_wanted": "0",
        "gas_used": "0",
        "events": [],
        "codespace": ""
      }
    ],
    "finalize_block_events": [
      {
        "attributes": [
          {
            "key": "",
            "value": "\n\u001e/cosmos.gov.v1.MsgVoteResponse",
            "index": false
          }
        ]
      }
    ],
    "validator_updates": [
      {
        "pub_key": {
          "Sum": {
            "type": "tendermint.crypto.PublicKey_Secp256K1",
            "value": {
              "secp256k1": "Cgl2YWxpZGF0b3ISMHRwdmFsb3BlcjFxOHFjdHV3bXZ1ZHB2bHZsNWNzdTNnd2E1emtnMmh6NTg4dnRsMxgB"
            }
          }
        }
      }
    ],
    "consensus_param_updates": null,
    "app_hash": null
  }
}
```

First, the `tx_results` entry is weird. The `data` field appears to be a base64 encoded string that decodes to `'\n\n\b\200\200�\n\020\200\216�\034\022\016\b�\215\006\022\004\b\200�\n\030\200\200@\032\t\n\aed25519'` which is not the correct content for that field. That field should be some protobuf encoding bytes and include `/cosmos.gov.v1.MsgVoteResponse`. The `ed25519` at the end suggests that it's related to the original`EndBlock.ConsensusParamUpdates.Validator` field. That `tx_results` entry also doesn't have any events, but should have 15.

The `finalize_block_events` are clearly wrong because it's only got one event that doesn't have either a type or key, neither of which should be possible. There should be 11 events there (all from the begin block).

### What did you expect to happen?

I expected all the block-results info for the requested block.

### How to reproduce it

Using a debugger and looking inside the `LoadFinalizeBlockResponse`, we got a `string(buf)` value. Then, I wrote a unit test that replicates the decoding behavior in `LoadFinalizeBlockResponse`. Specifically, I unmarshaled that `buf` value into a `ResponseFinalizeBlock`, expecting it to give an error, but it didn't.

In `LoadFinalizeBlockResponse`, it first attempts to convert the value (from the state store) into a new `ResponseFinalizeBlock`.  If that fails, it then tries to convert it to a `LegacyABCIResponses`. The problem is that sometimes the bytes from marshalling a `LegacyABCIResponses` can be unmarshaled as a `ResponseFinalizeBlock` without error. But since they're different objects, the resulting `ResponseFinalizeBlock` isn't even close to accurate.

I think part of the problem stems from the `Data` field. That field contains a protobuf encoded `sdk.TxMsgData` which just contains the response msg url(s). When trying to unmarshal the `ResponseFinalizeBlock`, those protobuf control bytes can end up being used as if they're part of the `ResponseFinalizeBlock` encoding.

### Values In Question

Value from state:
```
respBz :=[]byte("\n\xdc\x0e\x12\"\x12 \n\x1e/cosmos.gov.v1.MsgVoteResponse\x1a\xf7\x02[{\"msg_index\":0,\"events\":[{\"type\":\"message\",\"attributes\":[{\"key\":\"action\",\"value\":\"/cosmos.gov.v1.MsgVote\"},{\"key\":\"module\",\"value\":\"governance\"},{\"key\":\"sender\",\"value\":\"tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25\"}]},{\"type\":\"proposal_vote\",\"attributes\":[{\"key\":\"option\",\"value\":\"option:VOTE_OPTION_YES weight:\\\"1.000000000000000000\\\"\"},{\"key\":\"proposal_id\",\"value\":\"1\"}]}]}](\xc0\x9a\f0\x99\xe7\x04:`\n\ncoin_spent\x126\n\aspender\x12)tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25\x18\x01\x12\x1a\n\x06amount\x12\x0e381000000nhash\x18\x01:d\n\rcoin_received\x127\n\breceiver\x12)tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt\x18\x01\x12\x1a\n\x06amount\x12\x0e381000000nhash\x18\x01:\x97\x01\n\btransfer\x128\n\trecipient\x12)tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt\x18\x01\x125\n\x06sender\x12)tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25\x18\x01\x12\x1a\n\x06amount\x12\x0e381000000nhash\x18\x01:@\n\amessage\x125\n\x06sender\x12)tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25\x18\x01:\\\n\x02tx\x12\x1c\n\x03fee\x12\x1310000000000000nhash\x18\x01\x128\n\tfee_payer\x12)tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25\x18\x01:c\n\x02tx\x12#\n\x0fmin_fee_charged\x12\x0e381000000nhash\x18\x01\x128\n\tfee_payer\x12)tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25\x18\x01:>\n\x02tx\x128\n\aacc_seq\x12+tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25/2\x18\x01:m\n\x02tx\x12g\n\tsignature\x12X8J3dGzkMy8bz3ldIS/80nNgtLV2+A/0NU7RNbWpYy8RqNTJT/NLkXFe2OqFugwYl7QQVjT3iG1vzW8jhswSV1A==\x18\x01:-\n\amessage\x12\"\n\x06action\x12\x16/cosmos.gov.v1.MsgVote\x18\x01:e\n\rproposal_vote\x12@\n\x06option\x124option:VOTE_OPTION_YES weight:\"1.000000000000000000\"\x18\x01\x12\x12\n\vproposal_id\x12\x011\x18\x01:X\n\amessage\x12\x16\n\x06module\x12\ngovernance\x18\x01\x125\n\x06sender\x12)tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25\x18\x01:d\n\ncoin_spent\x126\n\aspender\x12)tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25\x18\x01\x12\x1e\n\x06amount\x12\x129999619000000nhash\x18\x01:h\n\rcoin_received\x127\n\breceiver\x12)tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt\x18\x01\x12\x1e\n\x06amount\x12\x129999619000000nhash\x18\x01:\x9b\x01\n\btransfer\x128\n\trecipient\x12)tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt\x18\x01\x125\n\x06sender\x12)tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25\x18\x01\x12\x1e\n\x06amount\x12\x129999619000000nhash\x18\x01:@\n\amessage\x125\n\x06sender\x12)tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25\x18\x01\x12)\x12'\n\n\b\x80\x80\xc0\n\x10\x80\x8e\xce\x1c\x12\x0e\b\xa0\x8d\x06\x12\x04\b\x80\xc6\n\x18\x80\x80@\x1a\t\n\aed25519\x1a\x8c\n\n@\n\amessage\x125\n\x06sender\x12)tp1m3h30wlvsf8llruxtpukdvsy0km2kum8j5f3gk\x18\x01\n\x8f\x01\n\x04mint\x12&\n\fbonded_ratio\x12\x140.000010000000000000\x18\x01\x12#\n\tinflation\x12\x140.000000000000000000\x18\x01\x12+\n\x11annual_provisions\x12\x140.000000000000000000\x18\x01\x12\r\n\x06amount\x12\x010\x18\x01\ne\n\ncoin_spent\x126\n\aspender\x12)tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt\x18\x01\x12\x1f\n\x06amount\x12\x1310000000000000nhash\x18\x01\ni\n\rcoin_received\x127\n\breceiver\x12)tp1jv65s3grqf6v6jl3dp4t6c9t9rk99cd8a624xf\x18\x01\x12\x1f\n\x06amount\x12\x1310000000000000nhash\x18\x01\n\x9c\x01\n\btransfer\x128\n\trecipient\x12)tp1jv65s3grqf6v6jl3dp4t6c9t9rk99cd8a624xf\x18\x01\x125\n\x06sender\x12)tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt\x18\x01\x12\x1f\n\x06amount\x12\x1310000000000000nhash\x18\x01\n@\n\amessage\x125\n\x06sender\x12)tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt\x18\x01\n\x84\x01\n\x0fproposer_reward\x120\n\x06amount\x12$500000000000.000000000000000000nhash\x18\x01\x12?\n\tvalidator\x120tpvaloper1q8qctuwmvudpvlvl5csu3gwa5zkg2hz588vtl3\x18\x01\n~\n\ncommission\x12/\n\x06amount\x12#50000000000.000000000000000000nhash\x18\x01\x12?\n\tvalidator\x120tpvaloper1q8qctuwmvudpvlvl5csu3gwa5zkg2hz588vtl3\x18\x01\n|\n\arewards\x120\n\x06amount\x12$500000000000.000000000000000000nhash\x18\x01\x12?\n\tvalidator\x120tpvaloper1q8qctuwmvudpvlvl5csu3gwa5zkg2hz588vtl3\x18\x01\n\x7f\n\ncommission\x120\n\x06amount\x12$930000000000.000000000000000000nhash\x18\x01\x12?\n\tvalidator\x120tpvaloper1q8qctuwmvudpvlvl5csu3gwa5zkg2hz588vtl3\x18\x01\n}\n\arewards\x121\n\x06amount\x12%9300000000000.000000000000000000nhash\x18\x01\x12?\n\tvalidator\x120tpvaloper1q8qctuwmvudpvlvl5csu3gwa5zkg2hz588vtl3\x18\x01")
```

The same thing declared as an object:
```go
    event := func(t string, attrs ...abci.EventAttribute) abci.Event {
        return abci.Event{Type: t, Attributes: attrs}
    }
    attr := func(key, value string) abci.EventAttribute {
        return abci.EventAttribute{Key: key, Value: value, Index: true}
    }

    respObj := &cmtstate.LegacyABCIResponses{
        DeliverTxs: []*abci.ExecTxResult{
            {
                Code:      0,
                Data:      []byte("\x12 \n\x1e/cosmos.gov.v1.MsgVoteResponse"),
                Log:       "[{\"msg_index\":0,\"events\":[{\"type\":\"message\",\"attributes\":[{\"key\":\"action\",\"value\":\"/cosmos.gov.v1.MsgVote\"},{\"key\":\"module\",\"value\":\"governance\"},{\"key\":\"sender\",\"value\":\"tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25\"}]},{\"type\":\"proposal_vote\",\"attributes\":[{\"key\":\"option\",\"value\":\"option:VOTE_OPTION_YES weight:\\\"1.000000000000000000\\\"\"},{\"key\":\"proposal_id\",\"value\":\"1\"}]}]}]",
                Info:      "",
                GasWanted: 200000,
                GasUsed:   78745,
                Codespace: "",
                Events: []abci.Event{
                    event("coin_spent",
                        attr("spender", "tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25"),
                        attr("amount", "381000000nhash")),
                    event("coin_received",
                        attr("receiver", "tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt"),
                        attr("amount", "381000000nhash")),
                    event("transfer",
                        attr("recipient", "tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt"),
                        attr("sender", "tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25"),
                        attr("amount", "381000000nhash")),
                    event("message", attr("sender", "tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25")),
                    event("tx",
                        attr("fee", "10000000000000nhash"),
                        attr("fee_payer", "tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25")),
                    event("tx",
                        attr("min_fee_charged", "381000000nhash"),
                        attr("fee_payer", "tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25")),
                    event("tx", attr("acc_seq", "tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25/2")),
                    event("tx", attr("signature", "8J3dGzkMy8bz3ldIS/80nNgtLV2+A/0NU7RNbWpYy8RqNTJT/NLkXFe2OqFugwYl7QQVjT3iG1vzW8jhswSV1A==")),
                    event("message", attr("action", "/cosmos.gov.v1.MsgVote")),
                    event("proposal_vote",
                        attr("option", "option:VOTE_OPTION_YES weight:\"1.000000000000000000\""),
                        attr("proposal_id", "1")),
                    event("message",
                        attr("module", "governance"),
                        attr("sender", "tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25")),
                    event("coin_spent",
                        attr("spender", "tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25"),
                        attr("amount", "9999619000000nhash")),
                    event("coin_received",
                        attr("receiver", "tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt"),
                        attr("amount", "9999619000000nhash")),
                    event("transfer",
                        attr("recipient", "tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt"),
                        attr("sender", "tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25"),
                        attr("amount", "9999619000000nhash")),
                    event("message", attr("sender", "tp1q8qctuwmvudpvlvl5csu3gwa5zkg2hz5cw3l25")),
                },
            },
        },
        EndBlock: &cmtstate.ResponseEndBlock{
            ValidatorUpdates: nil,
            ConsensusParamUpdates: &cmttypes.ConsensusParams{
                Block:     &cmttypes.BlockParams{MaxBytes: 22020096, MaxGas: 60000000},
                Evidence:  &cmttypes.EvidenceParams{MaxAgeNumBlocks: 100000, MaxAgeDuration: 172800000000000, MaxBytes: 1048576},
                Validator: &cmttypes.ValidatorParams{PubKeyTypes: []string{"ed25519"}},
                Version:   nil,
                Abci:      nil,
            },
            Events: nil,
        },
        BeginBlock: &cmtstate.ResponseBeginBlock{
            Events: []abci.Event{
                event("message", attr("sender", "tp1m3h30wlvsf8llruxtpukdvsy0km2kum8j5f3gk")),
                event("mint",
                    attr("bonded_ratio", "0.000010000000000000"),
                    attr("inflation", "0.000000000000000000"),
                    attr("annual_provisions", "0.000000000000000000"),
                    attr("amount", "0")),
                event("coin_spent",
                    attr("spender", "tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt"),
                    attr("amount", "10000000000000nhash")),
                event("coin_received",
                    attr("receiver", "tp1jv65s3grqf6v6jl3dp4t6c9t9rk99cd8a624xf"),
                    attr("amount", "10000000000000nhash")),
                event("transfer",
                    attr("recipient", "tp1jv65s3grqf6v6jl3dp4t6c9t9rk99cd8a624xf"),
                    attr("sender", "tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt"),
                    attr("amount", "10000000000000nhash")),
                event("message", attr("sender", "tp17xpfvakm2amg962yls6f84z3kell8c5l2udfyt")),
                event("proposer_reward",
                    attr("amount", "500000000000.000000000000000000nhash"),
                    attr("validator", "tpvaloper1q8qctuwmvudpvlvl5csu3gwa5zkg2hz588vtl3")),
                event("commission",
                    attr("amount", "50000000000.000000000000000000nhash"),
                    attr("validator", "tpvaloper1q8qctuwmvudpvlvl5csu3gwa5zkg2hz588vtl3")),
                event("rewards",
                    attr("amount", "500000000000.000000000000000000nhash"),
                    attr("validator", "tpvaloper1q8qctuwmvudpvlvl5csu3gwa5zkg2hz588vtl3")),
                event("commission",
                    attr("amount", "930000000000.000000000000000000nhash"),
                    attr("validator", "tpvaloper1q8qctuwmvudpvlvl5csu3gwa5zkg2hz588vtl3")),
                event("rewards",
                    attr("amount", "9300000000000.000000000000000000nhash"),
                    attr("validator", "tpvaloper1q8qctuwmvudpvlvl5csu3gwa5zkg2hz588vtl3")),
            },
        },
    }
```

### Anything else we need to know

You can see my unit test here: https://github.com/provenance-io/provenance/blob/dwedul/block-results/app/block_results_test.go
A run of those tests can be seen here: https://github.com/provenance-io/provenance/actions/runs/9687931502/job/26733554500

**Comment by melekes** [comment]
Could this https://github.com/cometbft/cometbft/issues/3002 be related? cc @andynog

**Comment by iramiller** [comment]
This issue appears to be related to #3002. 

It seems that the choice of deserialization types here should be based on a version number check from the block header instead of guess and check deserialization attempts. As noted the protobuf overlapping control chars and structures here can cause undefined behaviors with many unexpected edge and corner cases. Further there may eventually be more that two versions (current and legacy) in play in block history over time which breaks the fall back method of deserializing.

**Comment by SpicyLemon** [comment]
> Could this #3002 be related? cc @andynog

I'd be willing to bet that these two issues are the same bug.

**Comment by andynog** [comment]
Thanks @SpicyLemon, I believe it's the same issue. 

We have a [PR](https://github.com/cometbft/cometbft/pull/3031) with a fix that we are requesting people to test before we merge and backport for additional validation. 

Would you be able to cherry-pick these changes to a local branch and test ?

**Comment by SpicyLemon** [comment]
> We have a [PR](https://github.com/cometbft/cometbft/pull/3031) with a fix that we are requesting people to test before we merge and backport for additional validation.
> 
> Would you be able to cherry-pick these changes to a local branch and test ? Or would there be a better way for you to test this ?

tl;dr: That PR appears to resolve this issue.

Here's what I did (all locally on my Mac):

1. Cloned the CometBFT repo, and checked out `v0.38.9`.
2. Applied [the fix](https://github.com/cometbft/cometbft/pull/3031) to it.
3. Switched over to my clone of our SDK fork and checked out `v0.50.7-pio-1` (our most recent version) and added a replace line to `go.mod` to make it use my local version of cometbft.
4. Switched to my clone of our provenance repo and checked out `v1.18.0`. That's our current mainnet version, which uses `v0.34.29` of CometBFT (and `v0.46.13-pio-4` of the SDK).
5. Built that old version and started a new single-node chain using it.
6. Submitted a software upgrade proposal and voted it through.
7. Got the block results for the height with the vote msg/tx using `curl 'http://127.0.0.1:26657/block_results?height=14'
8. Waited for the node to reach the upgrade height.
9. Stopped the node.
10. Checked out `tag-v1.19.0-rc2` (of provenance). This is currently (as of Friday) what our testnet uses, which uses `v0.38.7` of CometBFT (and `v0.50.7-pio-1` of the SDK).
11. Updated the `go.mod` and used replace lines to use local versions of both the SDK and CometBFT.
12. Built the new version with the fix.
13. Started the node up again.
14. Got the block results for the height with the vote msg/tx using both `curl 'http://127.0.0.1:26657/block_results?height=14' and the CLI command `provenanced q block-results 14`.

The content of the block results query (after the upgrade) now looks right.

**Comment by melekes** [comment]
Thanks @SpicyLemon ❤️ 

Closing this issue in favor of https://github.com/cometbft/cometbft/issues/3002.
